src/main/java/com/google/devtools/build/lib/analysis/skylark/StarlarkAttributeTransitionProvider.java

```diff
@@ -39,17 +41,18 @@
 import java.util.List;
 
 /**
- * This class implements a {@link SplitTransitionProvider} to provide a starlark-defined transition
- * that rules can apply to their dependencies' configurations. This transition has access to (1) the
- * a map of the current configuration's build settings and (2) the configured attributes of the
- * given rule (not its dependencies').
+ * This class implements a {@link TransitionFactory} to provide a starlark-defined transition that
+ * rules can apply to their dependencies' configurations. This transition has access to (1) the a
+ * map of the current configuration's build settings and (2) the configured attributes of the given
+ * rule (not its dependencies').
  *
  * <p>For starlark defined rule class transitions, see {@link StarlarkRuleTransitionProvider}.
  *
  * <p>TODO(bazel-team): Consider allowing dependency-typed attributes to actually return providers
  * instead of just labels (see {@link SkylarkAttributesCollection#addAttribute}).
  */
-public class StarlarkAttributeTransitionProvider implements SplitTransitionProvider {
+public class StarlarkAttributeTransitionProvider
+    implements TransitionFactory<RuleTransitionData>, SplitTransitionProviderApi {
```

Comment:
nit: rename `SplitTransitionProviderApi` to `TransitionFactoryApi`. After all, the class is not a `SplitTransitionProvider` anymore.

Reply:
I'll leave it up to @juliexxia whether to rename, she was discussing a future plan to separate starlark patch and split transitions.

Reply:
This is confusing to me too. What's the purpose of SplitTransitionProviderApi? It looks like just an empty interface to expose SplitTransitionProvider to Starlark? And what does it mean to say isSplit() is true below?

My understanding is this may or may not be a split. So for the scenarios where it isn't a split what do those words mean?

(I guess this is mostly a question for @juliexxia).

Reply:
I think how it stands makes sense for this PR. I can always do clean up later. 

Currently all Starlark transitions attached to attributes (i.e., StarlarkAttributeTransitionProvider transitions) are natively treated like SplitTransitions. This isn't something that's exposed to the user (i.e. if you return a map instead of a map of maps, that works just fine), just something that was done so we didn't have to handle patch and split transitions here. 

SplitTransitionProviderApi was originally introduced for the benefit of exposing MultiArchSplitTransitionProvider to Starlark. That's it's only use case other than this. So SplitTransitionProviderApi is still technically the truth. And TransitionFactory seems to obscure the difference between split and patch transitions - @katre as a catch up question, are those both going to fall under this? I think it's mostly important in regards to the restriction that split transitions can't be attached directly to rules.

Reply:
Yes, TransitionFactory can produce both PatchTransition or SplitTransition. I had considered making it generic in which type it returns (so the signature would become TransitionFactory<PatchTransition, RuleTransitionData>), but
1. This ends up being cumbersome to use, and
2. The actual usage would be that Attribute has TransitionFactory<?, RuleTransitionData> so the difference isn't meaningful.

Actual code that cares can check the value of TransitionFactory.isSplit to see if the factory thinks it's returning a SplitTransition or not.